dotnet/src/Connectors/Connectors.GoogleVertexAI/Core/Gemini/Clients/GeminiChatCompletionClient.cs

```diff
@@ -0,0 +1,611 @@
+
+            state.AddLastMessageToChatHistoryAndRequest();
+            await this.ProcessFunctionsAsync(state, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    private static ChatCompletionState ValidateInputAndCreateChatCompletionState(
+        ChatHistory chatHistory,
+        Kernel? kernel,
+        PromptExecutionSettings? executionSettings)
+    {
+        var chatHistoryCopy = new ChatHistory(chatHistory);
+        ValidateAndPrepareChatHistory(chatHistoryCopy);
+
+        var geminiExecutionSettings = GeminiPromptExecutionSettings.FromExecutionSettings(executionSettings);
+        ValidateMaxTokens(geminiExecutionSettings.MaxTokens);
+
+        return new ChatCompletionState()
+        {
+            AutoInvoke = CheckAutoInvokeCondition(kernel, geminiExecutionSettings),
+            ChatHistory = chatHistory,
+            ExecutionSettings = geminiExecutionSettings,
+            GeminiRequest = CreateRequest(chatHistoryCopy, geminiExecutionSettings, kernel),
+            Kernel = kernel! // not null if auto-invoke is true
+        };
+    }
+
+    private IEnumerable<StreamingChatMessageContent> GetStreamingChatMessageContentsOrPopulateStateForToolCalling(
+        ChatCompletionState state,
+        Stream responseStream)
+    {
+        using var chatResponsesEnumerator = this.ProcessChatResponseStream(responseStream).GetEnumerator();
+        while (chatResponsesEnumerator.MoveNext())
+        {
+            var messageContent = chatResponsesEnumerator.Current!;
+            if (state.AutoInvoke && messageContent.ToolCalls is not null)
+            {
+                if (chatResponsesEnumerator.MoveNext())
+                {
+                    // We disable auto-invoke because we have more than one message in the stream.
+                    // This scenario should not happen but I leave it as a precaution
+                    state.AutoInvoke = false;
+                    chatResponsesEnumerator.Reset();
+                    continue;
+                }
+
+                // If function call was returned there is no more data in stream
+                state.LastMessage = messageContent;
+                yield break;
+            }
+
+            // We disable auto-invoke because the first message in the stream doesn't contain ToolCalls or auto-invoke is already false
+            state.AutoInvoke = false;
+
+            // If we don't want to attempt to invoke any functions, just return the result.
+            yield return this.GetStreamingChatContentFromChatContent(messageContent);
+        }
+    }
+
+    private async Task ProcessFunctionsAsync(ChatCompletionState state, CancellationToken cancellationToken)
+    {
+        this.Logger.LogDebug("Tool requests: {Requests}", state.LastMessage!.ToolCalls!.Count);
+        this.Logger.LogTrace("Function call requests: {FunctionCall}",
+            string.Join(", ", state.LastMessage.ToolCalls.Select(ftc => ftc.ToString())));
```

Comment:
This logging should be guarded by a check for IsEnabled. It's doing non-trivial work at the call site even if LogLevel.Trace isn't enabled. Same goes for any other similar call sites.

Reply:
Okay, I didn't know that not checking enabled LogLevel has performance impact.

Reply:
done